Shield.Communication/Services/Wifi.cs

```diff
@@ -0,0 +1,297 @@
+            }
+
+            // We do not have an output stream yet so create one.
+            try
+            {
+                IOutputStream outputStream = await listener.GetOutputStreamAsync(args.RemoteAddress, args.RemotePort);
+
+                // It might happen that the OnMessage was invoked more than once before the GetOutputStreamAsync completed.
+                // In this case we will end up with multiple streams - make sure we have just one of it.
+                RemotePeer peer;
+                lock (this)
+                {
+                    if (CoreApplication.Properties.TryGetValue("remotePeer", out outObj))
+                    {
+                        peer = (RemotePeer)outObj;
+                    }
+                    else
+                    {
+                        peer = new RemotePeer(outputStream, args.RemoteAddress, args.RemotePort);
+                        CoreApplication.Properties.Add("remotePeer", peer);
+                    }
+                }
+
+                EchoMessage(peer, args);
+            }
+            catch (Exception exception)
+            {
+                // If this is an unknown status it means that the error is fatal and retry will likely fail.
+                if (SocketError.GetStatus(exception.HResult) == SocketErrorStatus.Unknown)
+                {
+                    throw;
+                }
+
+               // NotifyUserFromAsyncThread("Connect failed with error: " + exception.Message, NotifyType.ErrorMessage);
+            }
+
+        }
+
+        async void EchoMessage(RemotePeer peer, DatagramSocketMessageReceivedEventArgs eventArguments)
+        {
+            if (!peer.IsMatching(eventArguments.RemoteAddress, eventArguments.RemotePort))
+            {
+                // In the sample we are communicating with just one peer. To communicate with multiple peers application
+                // should cache output streams (i.e. by using a hash map), because creating an output stream for each
+                //  received datagram is costly. Keep in mind though, that every cache requires logic to remove old
+                // or unused elements; otherwise cache turns into a memory leaking structure.
+                //NotifyUserFromAsyncThread(String.Format("Got datagram from {0}:{1}, but already 'connected' to {3}", eventArguments.RemoteAddress, eventArguments.RemotePort, peer), NotifyType.ErrorMessage);
+            }
+
+            try
+            {
+                var reader = eventArguments.GetDataReader();
+                var size = reader.UnconsumedBufferLength;
+                StringBuilder sb = new StringBuilder();
+                while (reader.UnconsumedBufferLength > 0)
+                {
+                    var b = reader.ReadByte();
+                    sb.Append((char) b);
+                }
+
+                var msg = sb.ToString();
+
+                if (msg.StartsWith("VS:"))
+                {
+                    msg = msg.Substring(3);
```

Comment:
possibly - this is solely for broadcast discovery up front. I specifically left-substringed it so that further data can be appended later with backward compatibility - and if it isn't preceded with what I check for, it is ignored. I think it's ok by design - happy to talk about it